File: protoqueue-compiler/src/main/java/net/port/transformer/compiler/common/Log.java
package net.port.transformer.compiler.common;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.tools.Diagnostic;

/**
 * Created by zhongyongsheng on 2018/4/14.
 */
public class Log {
    private Messager messager;
    private Element defaultElement;

    public Log(Messager messager, Element defaultElement) {
        this.messager = messager;
        this.defaultElement = defaultElement;
    }

    public void debug(Element element, String msg, Object... args) {
        messager.printMessage(Diagnostic.Kind.NOTE, safeFormat(msg, args), element);
    }

    public void debug(String msg, Object... args) {
        messager.printMessage(Diagnostic.Kind.NOTE, safeFormat(msg, args));
    }

    public void warn(Element element, String msg, Object... args) {
        messager.printMessage(Diagnostic.Kind.WARNING, safeFormat(msg, args), element);
    }

    public void warn(String msg, Object... args) {
        messager.printMessage(Diagnostic.Kind.WARNING, safeFormat(msg, args));
    }

    public void error(Element element, String msg, Object... args) {
        messager.printMessage(Diagnostic.Kind.ERROR, safeFormat(msg, args), element);
    }

    public void error(String msg, Object... args) {
        messager.printMessage(Diagnostic.Kind.ERROR, safeFormat(msg, args));
    }

    private String safeFormat(String msg, Object... args) {
        try {
            return String.format(msg, args);
        } catch (Exception e) {
            return msg;
        }
    }
}

File: protoqueue-compiler/src/main/java/net/protoqueue/compiler/common/Log.kt
package net.protoqueue.compiler.common

import javax.annotation.processing.Messager
import javax.lang.model.element.Element
import javax.tools.Diagnostic

/**
 * apt日志
 * Created by zhongyongsheng on 2018/4/14.
 */
class Log(private val messager: Messager) {

    fun debug(element: Element, msg: String, vararg args: Any) {
        messager.printMessage(Diagnostic.Kind.NOTE, safeFormat(msg, *args), element)
    }

    fun debug(msg: String, vararg args: Any) {
        messager.printMessage(Diagnostic.Kind.NOTE, safeFormat(msg, *args))
    }

    fun warn(element: Element, msg: String, vararg args: Any) {
        messager.printMessage(Diagnostic.Kind.WARNING, safeFormat(msg, *args), element)
    }

    fun warn(msg: String, vararg args: Any) {
        messager.printMessage(Diagnostic.Kind.WARNING, safeFormat(msg, *args))
    }

    fun error(element: Element, msg: String, vararg args: Any) {
        messager.printMessage(Diagnostic.Kind.ERROR, safeFormat(msg, *args), element)
    }

    fun error(msg: String, vararg args: Any) {
        messager.printMessage(Diagnostic.Kind.ERROR, safeFormat(msg, *args))
    }

    private fun safeFormat(msg: String, vararg args: Any): String {
        return try {
            String.format(msg, *args) + "\r\n" //加上日志才会能换行
        } catch (e: Exception) {
            msg
        }
    }
}
